File: internal/collector/clamav.go
package collector

import (
	"context"
	"log/slog"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/rknightion/opnsense-exporter/opnsense"
)

// clamavCollector surfaces ClamAV (os-clamav plugin) engine version and
// signature-database freshness. Running state is already covered by the
// services collector, so this stays focused on the data that lets an operator
// alert on "signatures older than N hours" (a broken/unreachable freshclam
// mirror is otherwise a silent failure mode).
type clamavCollector struct {
	log *slog.Logger

	info             *prometheus.Desc
	signaturesTotal  *prometheus.Desc
	dbVersion        *prometheus.Desc
	dbBuiltTimestamp *prometheus.Desc

	subsystem string
	instance  string
}

func init() {
	collectorInstances = append(collectorInstances, &clamavCollector{
		subsystem: ClamAVSubsystem,
	})
}

func (c *clamavCollector) Name() string {
	return c.subsystem
}

func (c *clamavCollector) Register(namespace, instanceLabel string, log *slog.Logger) {
	c.log = log
	c.instance = instanceLabel
	c.log.Debug("Registering collector", "collector", c.Name())

	c.info = buildPrometheusDesc(c.subsystem, "version_info",
		"ClamAV engine version information (value is always 1; use labels)",
		[]string{"version"})

	c.signaturesTotal = buildPrometheusDesc(c.subsystem, "signatures",
		"Total number of ClamAV virus signatures loaded across all databases (0 both when freshclam has never run and when the field is unavailable)",
		nil)

	c.dbVersion = buildPrometheusDesc(c.subsystem, "signature_db_version",
		"Version number of a ClamAV signature database (increments on every freshclam update), by database",
		[]string{"db"})

	c.dbBuiltTimestamp = buildPrometheusDesc(c.subsystem, "signature_db_built_timestamp_seconds",
		"Unix timestamp of when a ClamAV signature database was built, from the freshclam build metadata, by database. Absent when the build-date fragment could not be parsed.",
		[]string{"db"})
}

func (c *clamavCollector) Describe(ch chan<- *prometheus.Desc) {
	ch <- c.info
	ch <- c.signaturesTotal
	ch <- c.dbVersion
	ch <- c.dbBuiltTimestamp
}

func (c *clamavCollector) Update(ctx context.Context, client *opnsense.Client, ch chan<- prometheus.Metric) *opnsense.APICallError {
	data, err := client.FetchClamAVVersion()
	if err != nil {
		return err
	}

	// os-clamav absent → Present=false; stay silent (#204, mirrors the ACME/
	// SMART/DynDNS plugin-gated convention).
	if !data.Present {
		return nil
	}

	ch <- prometheus.MustNewConstMetric(c.info, prometheus.GaugeValue, 1, data.EngineVersion, c.instance)

	if data.HasTotalSignatures {
		ch <- prometheus.MustNewConstMetric(c.signaturesTotal, prometheus.GaugeValue, float64(data.TotalSignatures), c.instance)
	}

	for _, db := range data.Databases {
		ch <- prometheus.MustNewConstMetric(c.dbVersion, prometheus.GaugeValue, float64(db.Version), db.Name, c.instance)

		if db.BuiltTimestamp > 0 {
			ch <- prometheus.MustNewConstMetric(c.dbBuiltTimestamp, prometheus.GaugeValue, db.BuiltTimestamp, db.Name, c.instance)
		}
	}

	return nil
}

File: grafana/tabs/clamav.py
"""
ClamAV tab — covers all opnsense_clamav_* metrics.

The os-clamav plugin is default-on (--exporter.disable-clamav) but plugin-gated
(silent 404 when not installed), so the whole tab is gated on the presence of
opnsense_clamav_version_info. Running state is already covered by the Services
tab; this tab is purely about signature freshness — a stale/broken freshclam
mirror is otherwise a silent failure mode (#204).

Rows:
  1. Engine & Signatures — engine version info table; total signatures stat
  2. Signature Databases  — per-database (daily/main/bytecode) version bar
                             gauge and build-age table
"""

from builder import Builder, sel, epoch_ms


def build(b: Builder):
    # ---- Sentinels ---------------------------------------------------------
    b.sentinel("has_clamav", "label_values(opnsense_clamav_version_info, __name__)")

    # ================================================================
    # Row 1: Engine & Signatures
    # ================================================================
    engine_info = b.table(
        "ClamAV Engine Version",
        [sel("opnsense_clamav_version_info")],
        w=12, h=4,
        excludes=["Value", "__name__", "job", "instance"],
        renames={
            "version": "Engine Version",
            "opnsense_instance": "Instance",
        },
        desc="ClamAV (clamd) engine version (info metric — value is always 1; use labels).",
    )
    signatures_total = b.stat(
        "Total Signatures Loaded",
        sel("opnsense_clamav_signatures"),
        unit="short", w=6, h=4,
        thresholds=[{"color": "blue", "value": None}, {"color": "red", "value": 0}],
        desc=(
            "Total number of virus signatures loaded across all databases. "
            "0 both when freshclam has never run and when the field is unavailable."
        ),
    )

    # ================================================================
    # Row 2: Signature Databases
    # ================================================================
    db_version = b.bargauge(
        "Signature Database Versions",
        [(sel("opnsense_clamav_signature_db_version"), "{{db}}")],
        unit="short", w=8, h=8,
        desc="Numeric version of each signature database (daily/main/bytecode); increments on every freshclam update.",
    )
    db_age = b.table(
        "Signature Database Freshness",
        [f"(time() - {sel('opnsense_clamav_signature_db_built_timestamp_seconds')}) / 3600"],
        w=16, h=8,
        excludes=["__name__", "job", "instance"],
        renames={
            "db": "Database",
            "Value": "Age (hours)",
            "opnsense_instance": "Instance",
        },
        unit_overrides={"Age (hours)": "h"},
        sort_by="Age (hours)",
        sort_desc=True,
        desc=(
            "Hours since each signature database was last built by freshclam "
            "(sorted oldest first). A large or growing value indicates freshclam "
            "is failing or the mirror is unreachable — the whole point of this metric."
        ),
    )
    db_built = b.table(
        "Signature Database Built",
        [epoch_ms(sel("opnsense_clamav_signature_db_built_timestamp_seconds"))],
        w=8, h=8,
        excludes=["__name__", "job", "instance"],
        renames={
            "db": "Database",
            "Value": "Built",
            "opnsense_instance": "Instance",
        },
        unit_overrides={"Built": "dateTimeAsIso"},
        sort_by="Database",
        desc="Timestamp each signature database was built, from the freshclam build metadata.",
    )

    # ================================================================
    # Tab assembly
    # ================================================================
    b.tab("ClamAV", [
        b.row("Engine & Signatures",
              [engine_info, signatures_total],
              present="has_clamav"),
        b.row("Signature Databases",
              [db_version, db_age, db_built],
              present="has_clamav"),
    ])
